Change absolute tolerances

luciansmith opened the pull request:
Change how we use absolute tolerances to match Copasi:  multiply every element by the absolute tolerance, as sort of a poor man's relative tolerance.

Also a few other cleanup things, especially in the docs.

luciansmith committed:
First pass at setting absolute tolerances differently.

luciansmith committed:
Update minor version number.

The major version number should probably be changed, as this technically significantly changes behavior, but I'd like to save rr 3.0 to include some other larger changes as well.

luciansmith committed:
Change rules and add tests.

Change to how COPASI does things:  a single absolute tolerance value is multiplied by the values of the elements, and hence scaled to amounts when appropriate.

Remove the 'concentration tolerance' functions, as hopefully their removal will twig anyone who was using them to the new system.

luciansmith committed:
Fix azure-pipelines typo.

luciansmith committed:
Fix Euler docs, as an aside.

luciansmith committed:
Fix docs to match.

Also, we no longer have the ability to set integrator things with integrator.setting, so change all of that example code.

luciansmith committed:
Ensure that absolute_tolerance vector is always the right size.

There was no way to set the tolerance vector to the right size before changing the model, and when the model was changed, the tolerance vector was wrong.  So, just make sure it's right.

luciansmith committed:
Remove 'concentration tolerance' python tests.

luciansmith committed:
Remove 'concentration tolerance' python tests.

luciansmith committed:
Fix values.

The new tolerances make the algorithm slightly different.  But why would it ever have this many digits of precision?  Why would it ever work?  It mystifies me, but anyway, let's see what happens.

matthiaskoenig commented:
Hi Lucian, great seeing work on this.

In my opinion it would make a lot of sense to add SBML test cases to the test suite with small volumes. I.e. some simple integrations with models with volumes of 1E-12. This would ensure that simulators implement tolerances correct or at least give a red flag to people if integrators don't pass the small volume/tolerances tests in the SBML test suite.
I could imagine that some other tools have similar issues.

luciansmith commented:
That makes sense, but the purpose of the SBML Test Suite is supposed to be 'are you interpreting SBML correctly'; it's expressly *not* been supposed to test 'did you set the tolerances correctly' in the past.

However, perhaps this particular situation is worth testing there anyway!  If you file an issue on the SBML Test Suite issue tracker, I'll comment there with a summary of what I've discovered about the issue, too.

luciansmith committed:
More output for stochastic tests on Python.

luciansmith committed:
Update docs for C bindings.

I don't know why these are here and not in docs/.  Maybe they're in both places?  Anyway, no-longer-existing-functions now expunged.

luciansmith committed:
Try adding -s to pytest command.

luciansmith committed:
Try reducing the absolute tolerance, in case it's a speed thing.

luciansmith committed:
Destroy the N_Vector after creating it!
